src/main/java/com/wire/xenon/crypto/mls/CryptoMlsClient.kt

```diff
@@ -0,0 +1,41 @@
+package com.wire.xenon.crypto.mls
+
+import com.wire.crypto.client.ClientId
+import com.wire.crypto.client.CoreCryptoCentral
+import com.wire.crypto.client.MLSClient
+import com.wire.crypto.client.MLSGroupId
+import com.wire.crypto.client.MlsMessage
+import kotlinx.coroutines.runBlocking
+import java.io.Closeable
+import java.util.*
+
+class CryptoMlsClient : Closeable {
+    private var cryptoCentral: CoreCryptoCentral
+    private var mlsClient: MLSClient
+
+    constructor(clientId: String, clientDatabaseKey: String) {
+        runBlocking {
+            cryptoCentral = CoreCryptoCentral.invoke(
+                getDirectoryPath(clientId),
+                clientDatabaseKey)
+            mlsClient = cryptoCentral.mlsClient(ClientId(clientId))
+        }
+    }
+
+    private fun getDirectoryPath(clientId: String): String {
+        return "mls/$clientId"
+    }
+
+    // Group id needs to be fetched by API given qualified id of the conversation or maybe also create a cache/table, then take base64 string and byte array
+    // encryptedMessage is payload.data (string) -> decodeBase64Bytes (take the string as base64 and then get byte array)
+    fun decrypt(mlsGroupId: String, encryptedMessage: String): ByteArray? {
+        val mlsGroupIdBytes: ByteArray = Base64.getDecoder().decode(mlsGroupId)
+        val encryptedMessageBytes: ByteArray = Base64.getDecoder().decode(encryptedMessage)
+        val decryptedMessage = runBlocking { mlsClient.decryptMessage(MLSGroupId(mlsGroupIdBytes), MlsMessage(encryptedMessageBytes)) }
```

Comment:
I think the general docs of runBlocking kinda assume that you are in a coroutine environment, therefore it is telling us to not use runBlocking from inside a suspending function. Or you could have this problem if your Java/Kotlin app runs on async/reactive code (like Reactor).
Because in that case you will definetely block the thread.
I don't think we should have that problem, because we are running in a multi-threaded app (unlike an Android app)
My idea is that if you have 10 concurrent requests in the app, you have 10 threads, and "blocking" them to run the coroutine is the same as normal blocking code (I hope)

I was reading this article https://www.baeldung.com/kotlin/suspend-functions-from-java
I also tried with success to call the BuildersKt.runBlocking directly from Java, it was okay but I still needed a kotlin wrapper class because some functions were not available to be called from java (for another issue)

Reply:
This is just the theory, unsure if I'm correct, or how to possibly test it

Reply:
This is exactly the use case for `runBlocking` calls. It's being used correctly.

The docs say it is redundant to use `runBlocking` _from_ a suspend function. Not that it is redundant to use `runBlocking` when calling _to_ a suspend function.

------

❌ Scenario 1: Redundant. Can lead to thread starvation as one of the threads in the coroutine pool is now waiting for another coroutine.

```kotlin
suspend fun doLongCalculation() { /** do tons of work */ }

// Suspend
// ↓
suspend fun doWork() { // No sense in using runBlocking inside a suspend function
    runBlocking { doLongCalculation() }
}
```

☑️ Scenario 2: Not redundant. Perfectly valid. Thread will wait for the execution of the block inside

```kotlin
suspend fun doLongCalculation() { /** do tons of work */ }

// Not suspend. We want to bridge with non-suspendable context.
// ↓
fun main() { 
    runBlocking { doLongCalculation() }
}
```

------

But, there's a caveat when scaling.

If the codebase changes and there's a new coroutine that calls the function `main` on scenario 2, it's basically the same as scenario 1.

For example:

```kotlin
suspend fun doLongCalculation() { /** do tons of work */ }

// Not suspend. We want to bridge with non-suspendable context.
// ↓
fun doWork() { 
    runBlocking { doLongCalculation() }
}

// But, for some reason, some coroutines are calling doWork in another context:
suspend fun scheduleAndParalleliseWork() {
    coroutineScope {
        repeat(10) {
            launch {
                doWork() // Coroutine calling doWork, which uses runBlocking
            }
        }
    }
}
```

TL;DR: `runBlocking` is fine. `Coroutine -> runBlocking -> Coroutine` isn't.